wallet-api/src/Wallet/Emulator/Types.hs

```diff
@@ -0,0 +1,138 @@
+{-# LANGUAGE ConstraintKinds   #-}
+{-# LANGUAGE FlexibleContexts  #-}
+{-# LANGUAGE GADTs             #-}
+{-# LANGUAGE LambdaCase        #-}
+{-# LANGUAGE OverloadedStrings #-}
+module Wallet.Emulator.Types where
+
+import           Control.Monad.Except
+import           Control.Monad.Operational as Op
+import           Control.Monad.State
+import           Control.Monad.Writer
+import           Data.Either
+import           Data.Map                  as Map
+import           Data.Maybe
+import           Data.Text                 as T
+
+-- Basic types (should be replaced by "real" types later)
+
+-- don’t care what these are, some data to allow typeclasses
+data Txn = Txn Int
+    deriving (Show, Eq, Ord)
+-- each agent has its own wallet state
+data Agent = Agent Int
+    deriving (Show, Eq, Ord)
+
+type Block = [Txn]
+type Chain = [Block]
+type TxPool = [Txn]
+
+data Notification = BlockValidated Block
+                  | BlockHeight Int
+                  deriving (Show, Eq, Ord)
+
+-- Wallet code
+
+data WalletState = WalletState
+    deriving (Show, Eq, Ord)
+
+emptyWalletState :: WalletState
+emptyWalletState = WalletState
+
+-- manually records the list of transactions to be submitted
+newtype EmulatedWalletApi a = EmulatedWalletApi { runEmulatedWalletApi :: StateT WalletState (Writer [Txn]) a }
+    deriving (Functor, Applicative, Monad, MonadState WalletState, MonadWriter [Txn])
+
+-- TODO
+instance WalletAPI EmulatedWalletApi where
+    submitTxn txn = tell [txn]
+    handleNotifications = undefined
+
+-- TODO
+class WalletAPI m where
+    submitTxn :: Txn -> m ()
+    handleNotifications :: [Notification] -> m ()
+
+data FailureReason = FailureReason T.Text
+    deriving Show
+
+-- Emulator code
+
+-- This is so that assertions can tell whether and why things failed. They’re not visible through the wallet api, but only to assertions, which is helpful for testing
+data FailedTxn = FailedTxn Txn FailureReason
+    deriving Show
+type FailedTxPool = [FailedTxn]
+
+type Assertion = EmulatorState -> Maybe AssertionError
+data AssertionError = AssertionError T.Text
+    deriving Show
+
+isValidated :: Txn -> Assertion
+isValidated txn emState =
+    if notElem txn (join $ emChain emState) then Just $ AssertionError $ "Txn not validated: " <> T.pack (show txn) else Nothing
+
+-- | The type of events in the emulator. @n@ is the type (usually a monad) in which wallet actions
+-- take place.
+data Event n a where
+    -- these need the coordination code to tell us what to do
+    WalletAction :: Agent -> n () -> Event n [Txn]
+    -- these are always responded to in a routine way, check triggers, update wallet state etc.
+    WalletRecvNotification :: Agent -> [Notification] -> Event n [Txn]
+    BlockchainActions :: Event n Block
```

Comment:
Aren't blockchain actions the same as notifications (further up)? Or what other actions do you imagine the blockchain can take that are relevant for us here?

Reply:
Blockchain actions reflect actions that actually change the chain state. Notifications represent that information *reaching* wallets. Which is a distinct event that can happen later. Also "blockchain actions" here is an event that bundles up a bunch of core-node stuff that could maybe be distinct.